aten/src/ATen/native/MaskedBatchOps.cpp

```diff
@@ -0,0 +1,208 @@
+            .squeeze(-1);
+    dims = self[2].narrow(0, 0, 1);
+  } else if (d1 == 1 && d2 == 2) {
+    data = data.squeeze(-2);
+    mask =
+        at::bmm(self[1].narrow(1, 0, 1).unsqueeze(-2), other[1].narrow(1, 0, 1))
+            .squeeze(-2);
+    dims = other[2].narrow(0, 1, d2 - 1);
+  } else if (d1 == 2 and d2 == 2) {
+    mask = at::bmm(self[1].narrow(2, 0, 1), other[1].narrow(1, 0, 1));
+    dims = at::cat(std::vector<at::Tensor>(
+        {self[2].narrow(0, 0, 1), other[2].narrow(0, 1, d2 - 1)}));
+  } else {
+    throw std::runtime_error(
+        "matmul not implemented with batches of 3+D tensors");
+  }
+  return std::make_tuple(data, mask, dims);
+}
+
+// tensor shape operators
+
+std::tuple<Tensor, Tensor, Tensor> b_contiguous(TensorList self) {
+  if (self.size() != 3) {
+    throw std::invalid_argument(
+        "3 tensor in input TensorList expected, found " +
+        std::to_string(self.size()));
+  }
+  return std::make_tuple(self[0].contiguous(), self[1].contiguous(), self[2]);
+}
+
+// assumption: (sizes[i] == -1) === ith dimension is dynamic
+std::tuple<Tensor, Tensor, Tensor> b_view(TensorList self, IntList sizes_new) {
+  if (self.size() != 3) {
+    throw std::invalid_argument(
+        "3 tensor in input TensorList expected, found " +
+        std::to_string(self.size()));
+  }
+  int bs = self[0].size(0);
+  if (sizes_new[0] != 1 && sizes_new[0] != -1 && sizes_new[0] != bs)
+    throw std::runtime_error("first dim in view must be 1, -1, or batch size");
+  std::vector<long long> data_sizes = {bs}, mask_sizes = {bs};
+  Tensor dims_new = at::zeros(self[2].type(), (sizes_new.size() - 1));
+  for (int i = 1; i < int(sizes_new.size()); i++) {
+    data_sizes.push_back(sizes_new[i]);
+    mask_sizes.push_back(sizes_new[i] == -1 ? self[0].size(i) : 1);
+    dims_new[i - 1] = (sizes_new[i] == -1);
+  }
+  return std::make_tuple(
+      self[0].view(IntList(data_sizes)), self[1].view(IntList(mask_sizes)), dims_new);
+}
+
+// special operators
+
+std::tuple<Tensor, Tensor, Tensor> b_update(TensorList self, TensorList other) {
+  if (self.size() != 3) {
+    throw std::invalid_argument(
+        "3 tensor in input TensorList expected, found " +
+        std::to_string(self.size()));
+  }
+  if (other.size() != 3) {
+    throw std::invalid_argument(
+        "3 tensor in input 'other' TensorList expected, found " +
+        std::to_string(other.size()));
+  }
+  // TODO: maybe third arguments needed
+  auto update_mask = other[1];
+  auto data = at::where(update_mask, other[0], self[0]);
+  return std::make_tuple(data, update_mask, other[2]);
+}
+
+std::tuple<Tensor, Tensor, Tensor> b_synchronize(TensorList self) {
+  if (self.size() != 3) {
+    throw std::invalid_argument(
+        "3 tensor in input TensorList expected, found " +
+        std::to_string(self.size()));
+  }
+  auto mask_new = self[1] + (1 - self[1]);
+  return std::make_tuple(self[0], mask_new, self[2]);
+}
```

Comment:
Is there any reason why those can't be implemented in Python? It looks like they could be implemented in a way that would be completely traceable.

Reply:
@apaszke  we don't really have a reliable way of seamlessly calling Python functions (i.e. non-script and non-traced) from the script right now. @zdevito has proposed an `exemplar` system to be able to do this, but we haven't really committed to anything.

I also think it makes more sense to just implement these in C++ from a future-proofing standpoint anyway. We might want to emit custom CUDA kernels etc to do this batching stuff

Reply:
My point is that if you were to trace the body of those ops, then they would expand to a number of simpler ops, that would no longer need any special dispatch. It's basically a few masking ops + real operations, so there's nothing that we couldn't represent in the IR.

Reply:
Couldn't we treat those functions with `@script`? I'm not opposed to adding a few of those for experimental reasons, but it would be good to avoid 2x'ing the size of our interface just for this.